cip/1.accepted/CIP2021-07-07-Grouping-keys-and-aggregation-expressions.adoc

```diff
@@ -0,0 +1,833 @@
+
+While the simple rewrite semantics works nicely for the considered examples, it is limited.
+Specifically, it only supports aggregation expressions of the form `_agg_(_ex_)`.
+
+Cypher, however, also allows aggregation functions to appear as sub-expression of <ProjectionItem>s, i.e. Cypher allows <ProjectionItem>s with expressions of forms, such as
+
+* `_ex_~1~ + _agg_(_ex_~2~)`,
+* `_agg_(_ex_~1~) + _ex_~2~`, and
+* `_agg_~1~(_ex_~1~) + _ex~2~_ * _agg_~2~(_ex_~3~)`
+
+Such expressions can still be sensible and useful.
+
+.Aggregation functions as sub-expressions
+====
+The RETURN clause
+
+.[[Q4]]Q4
+[source, cypher]
+----
+RETURN a AS a, (a + SUM(b*c) - MIN(c)) * 2 AS foo
+----
+
+should produce a result that is grouped by `a` and `foo` should be computed for each group as the value `a` plus the sum of all products of `b` and `c` minus the smallest value of `c` multiplied by two, so that result is:
+
+.Result Q4
+|===
+|a|foo
+
+|1|32
+//(1 + (2*3 + 3*4) - 3) * 2
+|2|24
+//(2 + (3*5) - 5) * 2
+|===
+====
+
+[NOTE]
+====
+A less artificial example is calculating the total gross of an order as the discounted sum of the net values –– product price multiplied by amount –– of the order's line items in a query such as:
+
+[source, cypher]
+----
+MATCH
+(c:Customer)-[:LOCATED_IN]->(s:State),
+(c)-[:ORDERED]->(o:Order),
+(o)-[:INCLUDES]->(li:LineItem)-->(p:Product)
+RETURN s AS state, c AS customer, o AS order,
+       SUM(li.amount * p.price) * c.discount * s.vat AS totalGross
+----
+====
+
+It has been documented on multiple occasions (e.g. cf. http://opencypher.org/articles/2017/07/27/ocig1-aggregations-article/[First oCIG Meeting]) that the existing semantics of Cypher is imprecise on such queries.
+
+A precise semantics on such queries has to provide
+
+* A clear definition of which <ProjectionItem>s constitute the grouping keys
+* Clear rules of which sub-expressions are allowed in <ProjectionItem>s containing aggregation functions
+
+This proposal provides such a precise semantics.
+
+== Proposal
+
+=== Syntax
+
+This proposal does not propose any net-new syntax.
+
+=== Semantics
+
+The proposed grouping and aggregation semantics is defined as a rewrite to the baseline semantics (similar to <<Simple rewrite semantics>> discussed above).
+The proposed semantics does not cover all syntactically possible queries and hence requires a syntax restriction to prohibit queries that are not covered.
+We discuss the <<Rewrite>> and the <<Syntax restriction>> in the following two subsections.
+We simplify this discussion by ignoring row ordering and pagination as well as omitted aliases.
+Subsequently, we give a separate and brief consideration of how to these aspects fit into the proposal, cf. <<Row ordering and pagination>> and <<Omitted aliases>>, respectively.
+
+==== Rewrite
+
+For an expression _ex_, let _AGG_(_ex_) be the set of aggregating (sub-)expressions, i.e. (sub-)expressions _aggex_ of the form _agg_(_preEx_).
```

Comment:
It stands to a subexpression in _AGG_(_ex_). However, you are right in that the symbol is not needed since not used further down. So let's simply remove it:

```suggestion
For an expression _ex_, let _AGG_(_ex_) be the set of aggregating (sub-)expressions, i.e. (sub-)expressions of the form _agg_(_preEx_).
```